### PR TITLE
Update WIT and binary import/export syntax

### PR DESCRIPTION
**Pull request by alexcrichton**
This is an update to `WIT.md` and a minor update to `Binary.md` to reflect the changes discussed in #193. Lots more discussion is in the issue, and a summary of the changes is:

* WIT `use` statements are almost entirely replaced. Old `pkg` and `self`-style paths are all gone.
* WIT documents are gone as a concept, a package is a collection of interfaces/worlds now.
* The `externname` binary format production has been updated to either be a kebab-name or an "ID". The URL field has been dropped.
* Imports and exports of interfaces in WIT worlds is now done through IDs instead of kebab-names.

### CODE REVIEW

**Comment by alexcrichton** [comment]
I'll also note that I'm in the process of implementing these changes. Current work is:

* https://github.com/bytecodealliance/wasm-tools/pull/1027
* https://github.com/bytecodealliance/wasmtime/pull/6390

File: design/mvp/WIT.md [hunk 1]
```diff
@@ -28,6 +27,38 @@ package suitable for distribution.
...
 [components]: https://github.com/webassembly/component-model
 
+## Package identifiers
+
+All WIT packages are assigned an "ID". IDs look like `foo:bar@1.0` and have
+three components:
+
+* A namespace, for example `foo` in `foo:bar`. This namespace is intended to
+  disambiguate between registries, top-level organizations, etc. For example
+  WASI interfaces use the `wasi:` namespace.
```
Comment (lukewagner) [inline design/mvp/WIT.md:37]:
```suggestion
  WASI interfaces use the `wasi` namespace.
```

File: design/mvp/WIT.md [hunk 2]
```diff
@@ -28,6 +27,38 @@ package suitable for distribution.
...
+
+* A package name, for example `bar` in `foo:bar`. This provides a further
+  description of the contained functionality within the namespace. For example
```
Comment (lukewagner) [inline design/mvp/WIT.md:40]:
```suggestion
* A package name, for example `bar` in `foo:bar`. A package name groups together a set
  of interfaces and worlds that would otherwise be named with a common prefix. For example
```

File: design/mvp/WIT.md [hunk 3]
```diff
@@ -28,6 +27,38 @@ package suitable for distribution.
...
+  `wasi:clocks` has the package name `clocks`.
+
+* An optional version, specified as a major and minor number. Currently full
+  semver isn't allowed and instead only `1.0` is allowed, for example. Both
```
Comment (lukewagner) [inline design/mvp/WIT.md:44]:
```suggestion
* An optional version, specified as a major and minor number. Because WIT packages define
  *interfaces*, not *implementations*, there is not a patch number, as in full semver. Both
```

File: design/mvp/WIT.md [hunk 4]
```diff
@@ -62,8 +95,10 @@ An `interface` can contain [`use`][use] statements, [type][types] definitions,
 and [function][functions] definitions. For example:
 
 ```wit
-interface wasi-fs {
-  use pkg.types.{errno}
+package wasi:filesystem
+
+interface filesystem {
+  use types.{errno}
```
Comment (lukewagner) [inline design/mvp/WIT.md:101]:
```suggestion
package wasi:filesystem

interface types {
  use wasi:clocks.wall-clock.{datetime}
```
The goal here being to have this example slightly more-closely align with the real `wasi:filesystem` which wants to define the interface ID `wasi:filesystem/types`.  If this change makes sense, you'd probably stick a `datetime` inside the `stat` record and remove `errno` from the `result` of `stat-file`.

File: design/mvp/WIT.md [hunk 5]
```diff
@@ -137,217 +161,248 @@ Worlds can contain any number of imports and exports, and can be either a
...
   }
 }
 ```
 
-The name of the `import` or `export` is the name of the corresponding item in
-the final component. This can be different from the [`interface`][interfaces]
-name but must be a [valid identifier][identifiers].
+The kebab name of the `import` or `export` is the name of the corresponding item
+in the final component. This can be different from the [`interface`][interfaces]
+name but must be a [valid identifier][identifiers]. Note that kebab names cannot
```
Comment (lukewagner) [inline design/mvp/WIT.md:203]:
```suggestion
in the final component. Note that kebab names cannot
```

File: design/mvp/WIT.md [hunk 6]
```diff
@@ -137,217 +161,248 @@ Worlds can contain any number of imports and exports, and can be either a
...
-```
 
-If no `default` world is specified in the WIT document and no named world is
-explicitly chosen then bindings cannot be generated.
+world command {
+  // generates an import of the ID `wit:demo/my-interface`
+  import my-interface
+
+  // generates an import of the ID `wasi:random/random`
+  import wasi:random/random
```
Comment (lukewagner) [inline design/mvp/WIT.md:221]:
```suggestion
  // generates an import of the ID `wasi:filesystem/types`
  import wasi:filesystem/types
```

Comment (lukewagner) [reply design/mvp/WIT.md:221]: (To avoid the confusion of "why are there two `random`s?".)

File: design/mvp/WIT.md [hunk 7]
```diff
@@ -379,59 +436,25 @@ would generate this component:
...
   ))
 )
 ```
 
 Here it can be seen that despite the `world` only listing `host` as an import
-the component additionally imports a `shared` instance. This is due to the fact
-that the `use { ... } from shared` implicitly requires that `shared` is imported
-into the component as well.
+the component additionally imports a `wit:demo/shared` interface. This is due
+to the fact that the `use { ... } from shared` implicitly requires that
```
Comment (lukewagner) [inline design/mvp/WIT.md:453]:
```suggestion
to the fact that the `use shared.{ ... }` implicitly requires that
```

File: design/mvp/WIT.md [hunk 8]
```diff
@@ -1114,22 +1188,25 @@ would correspond to:
...
 (component
-  (type (export "host") (component
+  (type $demo (component
     (type $console (instance
       (export "log" (func (param "arg" string)))
     ))
-    (export "console" (instance (type $console)))
+    (export (interface "wit:demo/console") (instance (type $console)))
   ))
+  (export (interface "wit:demo/wit") (type $demo))
```
Comment (lukewagner) [inline design/mvp/WIT.md:1197]: I'm a bit confused where the final `/wit` comes at the end of `wit:demo/wit`.  If we want to preserve file grouping, I'd expect the ID to be `wit:demo/host` or even just the kebab-name `host` (highlighting the fact that `host` is not an *interface*).  If we don't want to preserve file grouping, then maybe export the whole package with the kebab-name of the package (again, because it's not an *interface*).

**Review by lukewagner — COMMENTED** [review]
Looks good!  Some small suggestions and one interesting question at the end about the encoding of Wit into Wat.

File: design/mvp/WIT.md (hunk 8 above)
Comment (alexcrichton) [reply design/mvp/WIT.md:1197]: Good question! This is here though not for interfaces but instead for the package itself. I copy-pasted this output from `wasm-tools component wit` which is why it's here, but my thinking is that the package name should be recorded somewhere, probably around here, but IDs require a `/foo` and when naming the package there's nothing to put there, so I arbitrarily put in `/wit`.

This doesn't actually reflect file structure though, it's only there to satisfy the validation requirements of an interface name being present. All contents of the package are in the component type itself.

Basically this ideally would be simply `"wit:demo"`, but that doesn't validate.

Comment (lukewagner) [reply design/mvp/WIT.md:1197]: Ah, hah, gotcha.  One observation is that, since the namespace, package name and version are already present inside each exported interface and world ID, there's no additional info provided by this.  In fact, I wonder if we even need this enclosing component-type at all: could the outermost component simply contain an *unnamed* `type`-definition for each interface-or-world which is a component-type that imports `use`d dependencies and exports the interface-or-world, establishing the ID via `(export (interface ...) ...)`?

Comment (alexcrichton) [reply design/mvp/WIT.md:1197]: Validation-wise, yes, that should work. Integration-wise though that would be more difficult I think though because index spaces are generally considered to be implementation details of wasms where the exports/imports define how the embedder interacts with it. 

One other issue though is that if a package is empty with no other interfaces or worlds I don't think the ID would be recorded anywhere?

File: design/mvp/WIT.md [hunk 9]
```diff
@@ -137,217 +161,248 @@ Worlds can contain any number of imports and exports, and can be either a
...
 
 world your-world {
-  import out-of-line: self.out-of-line
+  import out-of-line
   // ... is roughly equivalent to ...
-  import out-of-line2: interface {
+  import out-of-line: interface {
     the-function: func()
   }
 }
```
Comment (guybedford) [inline design/mvp/WIT.md:198]: I get this framing, and it's a nice approach, but I wasn't aware we were going to entirely separate kebab names as no longer being able to reference other dependencies at all.

Again for the componentization use case, if I want to target a dependency that is another component wasm file, it would be nice to be able to have tooling that could do eg:

```
world parent-world {
  import dep
}
```

and then be able to define `dep` to a local component when componentizing:

```
jco componentize x.js --world parent-world --dep dep=./child.wasm
```

It would be nice to leave the door open to interesting use cases like the above.

**Review by guybedford — COMMENTED** [review]
Really great work. This all seems in line with previous discussions. I have two major concerns:

1. Making package declarations required adds possible redundancy to componentization use cases.
2. Disallowing non-package imports as being resolved by tooling bypasses some interesting usage scenarios.

Both seem relatively surface-level restrictions to me though, and I've tried to go into as much detail as I can on these.

File: design/mvp/WIT.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukewagner) [reply design/mvp/WIT.md:37]: Instead of making namespaces optional, which I think would be a half-way measure, I think these cases should just not use `(interface <id>)` and produce plain kebab-names instead.

File: design/mvp/WIT.md [hunk 10]
```diff
@@ -28,6 +27,38 @@ package suitable for distribution.
 [IDL]: https://en.wikipedia.org/wiki/Interface_description_language
 [components]: https://github.com/webassembly/component-model
 
+## Package identifiers
+
+All WIT packages are assigned an "ID". IDs look like `foo:bar@1.0` and have
+three components:
```
[the remark this answers is not in the document]
Comment (lukewagner) [reply design/mvp/WIT.md:33]: I agree that it's useful to have `package` names be optional, in which case all the interfaces being defined by the collection of Wit documents passed to the tool don't have any associated interface ID: it's all kebab names.

File: design/mvp/WIT.md (hunk 8 above)
Comment (lukewagner) [reply design/mvp/WIT.md:1197]: I suppose another slight variation is for each interface/world to be a separate type definition that is exported with the ID (which would then be always be duplicated).  I guess mainly I was realizing that it doesn't seem like the component-types representing each individual interface/world need to be wrapped up into a single component type.

It's true that if the package was empty, there wouldn't be any record of the package name, but it seems like that's fine, since there's nothing to refer to.

Comment (alexcrichton) [reply design/mvp/WIT.md:1197]: I did actually initially go for the containerless format like you're mentioning. The names actually already duplicate the package name as each interface is exported with its full ID rather than just its name. The problem is that without a container you're generating an actual component with actual imports since interfaces may need to import interfaces they depend on. That would change things radically since this is supposed to be "just" type-level things.

File: design/mvp/WIT.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (alexcrichton) [reply design/mvp/WIT.md:33]: I agree it would be useful to be able to omit `package` altogether, but that raises enough questions for me that I'm hoping we can get by for now by not implementing it. For example what @lukewagner is proposing means that worlds have drastically different meanings whether or not the enclosing package has an ID or not, and I'm not sure I'm comfortable with that. I would expect some form of opt-in to "yes please interpret everything differently here" more than leaving out `package ...`.

For the JS componentization use case though what you'll be doing is inferring a `world` from a concrete component, and in such a situation there may not have actually ever been a `world` written down in WIT anywhere corresponding to what you're inferring. For example you're inferring a "minimized" version which will be missing functions that aren't otherwise required (e.g. you import `wasi:filesystem/filesystem`'s `stat` function but not `readdir` or something like that). The WIT here, however, entirely synthesized since all you're given was a component. Currently this synthesizes a dummy `package root:root` with a `world root { ... }` in the `Resolve` data structure, since otherwise there's no names to fill in.

Put another way I think this would be good to get to at some point but for now I'm hoping that writing `package foo:foo` at the top or something like that is "good enough".

File: design/mvp/WIT.md [hunk 11]
```diff
@@ -28,6 +27,38 @@ package suitable for distribution.
...
+Package identifiers are specified at the top of a WIT file via a `package`
+declaration:
+
+```wit
+package wasi:clocks
+```
+
+WIT packages can be defined in a collection of files and at least one of them
+must specify a `package` identifier. Multiple files can specify a `package` and
+they must all agree on what the identifier is.
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply design/mvp/WIT.md:63]: I'm tempted to start conservatively and require them in the text, but do you have a use case in mind where this would be helpful to have?

File: design/mvp/WIT.md (hunk 9 above)
Comment (guybedford) [reply design/mvp/WIT.md:205]: > If we go via the route described above where this form automatically uses kebab-names in some situations, I think that would satisfy the use case here?

Can you clarify what you mean by this?

File: design/mvp/WIT.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (guybedford) [reply design/mvp/WIT.md:63]: If we supported the kebab name variant `package foo` that would be fine, but in particular it would be nice to retain the ability to have WIT files that are not part of packages when just using their interfaces and inlining their types, without inlining their names. Ie at the top-level WIT authoring level where you don't even think in terms of a package name yet.

File: design/mvp/WIT.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (guybedford) [reply design/mvp/WIT.md:33]: @alexcrichton in the above case I would have expected that `wasi:filesystem/filesystem` would still be a direct import of the WASI externname host. It would only be the kebab name types (and implementations) which would be inlined into the component surely?

The fact that we are inferring a `world` from a component is exactly the use case though - componentization is creating components from worlds. Handling that in dependencies is exactly recreating worlds from **these** components. I don't think this use case should be surprising?

I don't fully follow how kebab names break the model in terms of them just being package-local resolutions. I'm fine with some kind of `root:root` pattern to make this stuff work internally, and I understand if you need to move ahead here, my main point is just asking are just:

1. Will these use cases be possible, so we are not cutting them out entirely?
2. Are we not making unnecessary restrictions now that will hinder these use cases unnecessarily now or in future?

[the remark this answers is not in the document]
Comment (guybedford) [reply design/mvp/WIT.md:33]: After discussing this further with @alexcrichton, he clarified there may still be ways of supporting kebab name external references in future. But for now that is a secondary use case, since inline types solve the majority of the problem. So effectively:

1. Yes these use cases are possible, but would require further design work in future motivated by real need
2. There are no restrictions being made now that inhibit the externalization of type definitions for kebab names in future

File: design/mvp/WIT.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (guybedford) [reply design/mvp/WIT.md:63]: After further discussion, the `local:pkg` convention for now will be fine. In future out-of-band configurations eg for kebab names may be possible, depending on how type externalization for them is managed.

**Review by guybedford — APPROVED** [review]
Thanks @alexcrichton for working through my questions today. To summarize -

The kebab names are still being supported fine with inline types which mostly handles the major use cases. Any additional use cases that may be supported in future for kebab names or instance imports can be designed on their own merits.

Finally, with regards to having good conventions for unknown package namespaces in the "optional" scenarios, I've suggested the use of `local:` in the examples as a possible name here for now. I'm open to other alternatives, but that would mitigate my concern of `wit:` being used as an example here then being taken up more widely (or some other ambiguous namespacing). Happy to bikeshed further of course!

File: design/mvp/WIT.md (hunk 8 above)
Comment (lukewagner) [reply design/mvp/WIT.md]: Good point, I was misreading the current output and you're right.  Thinking about it some more, given the following Wit:
```wit
package wasi:http
interface types {
  resource request { ... }
  resource response { ... }
}
interface handler {
  use types.{request, response}
  handle: func(request) -> response
}
world proxy {
  import handler
  export handler
}
```
I think perhaps we'd want to get the following Wat encoding:
```wat
(component
  (component (export "types")
    (type (export (interface "wasi:http/types")) (instance
      (export "request" (type (sub resource)))
      (export "response" (type (sub resource)))
    ))
  )
  (component (export "handler")
    (import (interface "wasi:http/types") (instance $types
      (export "request" (type (sub resource)))
      (export "response" (type (sub resource)))
    ))
    (alias export $types "request" (type $request))
    (alias export $types "response" (type $response))
    (type (export (interface "wasi:http/handler")) (instance
      (export "handle" (func (param "request" (own $request)) (result (own $response))))
    ))
  )
  (component (export "proxy")
    (type (export (interface "wasi:http/proxy")) (component
      (import (interface "wasi:http/types") (instance $types
        (export "request" (type (sub resource)))
        (export "response" (type (sub resource)))
      ))
      (alias export $types "request" (type $request))
      (alias export $types "response" (type $response))
      (import (interface "wasi:http/handler") (instance
        (export "handle" (func (param "request" (own $request)) (result (own $response))))
      ))
      (export (interface "wasi:http/handler") (instance
        (export "handle" (func (param "request" (own $request)) (result (own $response))))
      ))
    ))
  ))
)
```
Here, each interface and world is individually wrapped in a component that is able to express `use` dependencies via imports.  For worlds, there's currently never a need to have imports in the wrapper, but I'm not 100% sure this would never make sense in the future and it seems nice to have interfaces and worlds be symmetric, distinguished by whether the wrapper component exports an `instance` type (for an interface) or a `component` type (for a world).

A nice thing about this encoding is that it makes it clear that each interface/world definition is a standalone unit and thus packages can be arbitrarily split and merged.  It also makes clear that each interface's `use`s can be resolved independently and/or in multiple ways, depending on context, which is less clear with the current encoding where they alias each other directly.

Another nice thing is that, given an interface ID like `wasi:http/handler`, you can think of the `wasi:http` as mapping to the above component package and then the trailing `handler` projects out the `handler` subcomponent which is also how nested component *implementation* packages would work.  This also naturally generalizes to deeper nesting (i.e., `foo:bar/baz/quux`) if we ever wanted that.  Also, if we ever wanted to allow `foo:bar` to refer to a specific interface or world, we could achieve that by having the outermost component contain `(type (export (interface "foo:bar")) ...)`, which we could do while still allowing there to be interfaces/worlds named `foo:bar/baz`.  (E.g., `wasi:http` could map to what's currently called `wasi:http/proxy`, if we wanted.)

WDYT?

Comment (alexcrichton) [reply design/mvp/WIT.md]: I suspect something along these lines could work, but there's two issues given the current translation as I understand it:

1. An empty wit package would be translated as `(component)` which doesn't include the package's name and it's also ambiguous with an acutal empty component itself. While it's sort of an edge case this is all so low-level I'd ideally prefer to avoid any accidental ambiguities with things like this.
2. Do exports have to be of actual concrete components? An operation supported today is "extract the WIT from this concrete component" and that transparently works with actual components and wasm-encoded WIT packages. The detection is currently "does this component export a component type" since no real component will realistically do that for a long time. This detection, however, gets more complicated as the structure of the component looks "real" until you get to the exported instance/component types. Additionally I personally thought it was a nice feature to encode the entire WIT package in the type section, where this is sort of half actual runtime things half type things.

Comment (lukewagner) [reply design/mvp/WIT.md]: Yeah, both great questions.  I'm not sure these are the absolute best answers, but I'll put them out there to see what you think:
1. In a full workflow where this package gets published, the registry to which it was published would establish the name.  We've also talked about a registry-defined package metadata custom section that would contain information like this too.  I suppose why it seems fine to have a different answer here is that packages are not "semantic" things with IDs in the same way that interfaces and worlds are.
2. I could see s/component def/component type/ working too, almost equivalently, so I've flip-flopped in my head on this one.  For rendering `.wit` from `.wasm`, I think the "signal" would be any `(export (interface ...) (type ...))` and so the Wit-rendering tool would recurse over the component tree and splat out each of these it found into the generated `.wit` document.  As a nice side effect, you could then bundle `.wit` with *any* component `.wasm`.  The other small neat thing about using component defs as the "wrapper" is that last little thing I mentioned about being able to have the outermost component serve as the wrapper (possibly, at some point in the future).

File: design/mvp/WIT.md [hunk 12]
```diff
@@ -28,6 +27,44 @@ package suitable for distribution.
...
+  disambiguate between registries, top-level organizations, etc. For example
+  WASI interfaces use the `wasi` namespace.
+
+* A package name, for example `clocks` in `wasi:clocks`. A package name groups
+  together a set of interfaces and worlds that would otherwise be named with a
+  common prefix.
+
+* An optional version, specified as a major and minor number. Because WIT
+  packages define *interfaces*, not *implementations*, there is not a patch
+  number, as in full semver. Both version numbers must be unsigned integers.
```
Comment (lukewagner) [inline design/mvp/WIT.md:45]: I did realize a potential use case for including an optional patch version: when you're iterating on a pre-1.0 interface, and minor version bumps can be breaking, then you might want to use patch version changes to signal non-breaking interface changes.  Concretely, I could imagine that for Preview 2, we'd want to publish `0.` versions of WASI interfaces (say, `0.2`), and then if we make non-breaking changes post-Preview 2, we could publish them as `0.2.x` versions (saving `0.3` for Preview 3).

Comment (rylev) [reply design/mvp/WIT.md:45]: We could also allow a pre-release version number and potentially a build number. I think this would more closely match the actual semantics (i.e., it's not a patch; it's a pre-release)

Comment (tschneidereit) [reply design/mvp/WIT.md:45]: In the past we had discussed using patch versions for non-functional changes. Perhaps the only, but still important, change of this kind would be documentation changes. I think those warrant the inclusion of patch version numbers

File: design/mvp/WIT.md [hunk 13]
```diff
@@ -137,217 +168,247 @@ Worlds can contain any number of imports and exports, and can be either a
...
     the-function: func()
   }
 }
 ```
 
-The name of the `import` or `export` is the name of the corresponding item in
-the final component. This can be different from the [`interface`][interfaces]
-name but must be a [valid identifier][identifiers].
+The kebab name of the `import` or `export` is the name of the corresponding item
+in the final component. Note that kebab names cannot overlap and must be unique,
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply design/mvp/WIT.md:209]: Good point! I've clarified that this use case is allowed.

File: design/mvp/WIT.md [hunk 14]
```diff
@@ -28,6 +27,44 @@ package suitable for distribution.
...
+
+or
+
+```wit
+package wasi:clocks@1.2
+```
+
+WIT packages can be defined in a collection of files and at least one of them
+must specify a `package` identifier. Multiple files can specify a `package` and
+they must all agree on what the identifier is.
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply design/mvp/WIT.md:63]: @guybedford do you have thoughts on this?

File: design/mvp/WIT.md [hunk 15]
```diff
@@ -40,6 +77,8 @@ belong to an interface.
 An example of an interface is:
 
 ```wit
+package local:demo
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply design/mvp/WIT.md:80]: @guybedford you might have thoughts on this? (I don't mind either way myself)

File: design/mvp/WIT.md (hunk 12 above)
Comment (alexcrichton) [reply design/mvp/WIT.md]: Ok sounds reasonable to me, I went ahead and added full semver in including pre-release/build metadata.

File: design/mvp/WIT.md (hunk 8 above)
Comment (alexcrichton) [reply design/mvp/WIT.md]: Given how low-level this tooling is though I'd ideally prefer to not require or assume a registry in place. Basically otherwise everything will need to carry the name around manually separate from the wasm itself, whereas otherwise it's stored inline in the wasm with the current proposal.

I'm still not sure I understand the "neat" part here though myself. Everything feels far enough in the future that nothing is really solid enough that we would be able to predict that an alternate design would indeed fit neatly into a design one way or another. I'm hesitant to change designs now radically for a proposed future design which isn't fully fleshed out.

Comment (lukewagner) [reply design/mvp/WIT.md]: Regarding the registry aspect: sorry, I didn't mean to imply that the package name lives in the registry, but rather a custom section in the component binary that could also hold other package-metadata types of fields.

That being said, in the short term, maybe it's better to just make the minimal change to the existing encoding to unblock other stuff.  Because we'll eventually be publishing these Wit packages to a registry, the encoding scheme will need to be effectively standardized to allow tools to interoperate and then it'll be important to think long-term about some of these packaging questions, but we're not there yet.

**Review by lukewagner — APPROVED** [review]
lgtm, thanks!  Explainer.md and the text format still need to be updated, but I'm happy to do that as a quick follow-on once you merge this PR.

File: design/mvp/WIT.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (guybedford) [reply design/mvp/WIT.md:77]: My concern was exactly that asking users to choose an arbitrary namespace at an early stage risks exactly the naming problem namespaces are supposed to avoid in the first place.

File: design/mvp/WIT.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (guybedford) [reply design/mvp/WIT.md:60]: I'd be open to changing the constraint to a single file, but see no problem with the redundancy either.

[the remark this answers is not in the document]
Comment (guybedford) [reply design/mvp/WIT.md:60]: Also the concept of `main` doesn't quite apply to WIT currently since all WIT files are equal. Unless we introduce preferential lookup rules for default worlds and interfaces.

[the remark this answers is not in the document]
Comment (rylev) [reply design/mvp/WIT.md:60]: There are other tools that have this convention. For example, terraform has the convention of main files that (I believe) have no significance other than to say to the reader "look here first".

File: design/mvp/WIT.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (rylev) [reply design/mvp/WIT.md:77]: So does `local` actually have a meaning here or is this a convention you're trying to establish or is this purely descriptive? Either way, it would be nice to explicitly state that in this document.

**Comment by guybedford** [comment]
With regards to the ESM integration, it would be great to handle this in this PR as well since those behaviours do need to be worked out as part of this. My suggestion would be a note along the lines of the following bullet points:

- When importing components in the ESM integration, web-compatible components are those that do not use namespaces but instead use kebab names in the outer import and export positions, or optionally URLs in the import position.
- When converting components into web compatible components, namespaced exports may be provided by convention as interfaces nested under the namespace and package interface names. In addition, if there are no conflicts against other exports, the interface and package names may be further flattened without nesting in this convention.

**Comment by alexcrichton** [comment]
Ok I'm going to go ahead and merge this, if at least to have the current spec reflect what's implemented now in tooling. I talked with @lukewagner and he'll update `Explainer.md` shortly afterwards and I think any further minor updates can happen in subsequent PRs. Thanks all for the discussion here!

**Comment by lukewagner** [comment]
ESM-integration updates aren't "minor", but I do think this and other language integrations questions can be done as follow-ups without changing what's in the AST and binary, so it's a good idea to merge now and iterate forward.  I'll be working to update some other parts of Explainer.md shortly.
